### ISSUE TITLE
AOTAutograd exposes a runtime Segfault

### ISSUE THREAD

**Issue by sgligorijevicTT** [tenstorrent/tt-xla#4053]
Repro steps:
```
git checkout sgligorijevic/aotautograd-on-by-default
cmake --build build
pytest -svv tests/torch/training/test_basic.py::test_qwen3_multichip_backward
```

Expected behavior:
Either passes completely, or correctly fails with bad PCC
Actual behavior:
SEGFAULT

I noticed this is a single chip graph without inputs that gets executed after a multichip graph that faults, during device reopening.

**Comment by acolicTT** [comment]
I could't repro python segfaults in both debug and release builds (I rebased sgligorijevic/aotautograd-on-by-default on main yesterday). 
Instead, each time I have an error: ERR| Device count mismatch: 2 vs 1

This happens after the last graph execution, which has different mesh shape (1, 1) then previous ones (1, 2):

```
...
...      client_instance.cc:534      1| ClientInstance::getOrCreateMeshDevice - reshaping mesh device - [1, 2] -> [1, 1]
...          tensor_pool.cc:91       1| Moving tensors to host.
...      client_instance.cc:608      1| Closing parent mesh.
...      client_instance.cc:574      1| ClientInstance::openMeshDevice - setting fabric config: DISABLED
... loaded_executable_insta:266      1| LoadedExecutableInstance::PJRT_LoadedExecutable_GetExecutable
... loaded_executable_insta:286      1| LoadedExecutableInstance::PJRT_LoadedExecutable_AddressableDevices
...               stubs.inc:80    WARN| STUB: PJRT_Executable_GetCompiledMemoryStats
...       error_instance.cc:52       1| ErrorInstance::PJRT_Error_Message
...       error_instance.cc:61       1| ErrorInstance::PJRT_Error_GetCode
...       error_instance.cc:46       1| ErrorInstance::PJRT_Error_Destroy
...  executable_instance.cc:116      1| ExecutableInstance::PJRT_Executable_OptimizedProgram
...  executable_instance.cc:116      1| ExecutableInstance::PJRT_Executable_OptimizedProgram
...  executable_instance.cc:116      1| ExecutableInstance::PJRT_Executable_OptimizedProgram
...  executable_instance.cc:116      1| ExecutableInstance::PJRT_Executable_OptimizedProgram
...  executable_instance.cc:254      1| ExecutableInstance::PJRT_Executable_Serialize
...   device_description.cc:67       1| DeviceDescription::PJRT_DeviceDescription_Id
...   device_description.cc:67       1| DeviceDescription::PJRT_DeviceDescription_Id
...  executable_instance.cc:154      1| ExecutableInstance::PJRT_Executable_NumOutputs
... loaded_executable_insta:335      1| LoadedExecutableInstance::PJRT_LoadedExecutable_Execute
... flatbuffer_loaded_execu:192      1| FlatbufferLoadedExecutableInstance::Execute
... flatbuffer_loaded_execu:197    ERR| Device count mismatch: 2 vs 1
... 
```

It looks that there is a bug with sharding info propagation between graphs.

**Comment by jameszianxuTT** [comment]
I see something similar to @acolicTT, logs here: https://github.com/jameszianxuTT/logdump/blob/main/qwen_bw_fail_small_2026-04-23_16-21-24.log

Some small degenerate graph is being executed, with no inputs and producing some fake outputs. This looks like static cache + logits or something like that but filled with dummy values. I wonder what this is for and if it shows up on main (checking)

edit: Claude says this is gradient initialization which seems plausible given the output shapes and values

```
2026-04-23 16:20:46.246 (  22.683s) [        4F4F76C0]      module_builder.cc:1220     1| MLIR Module shlo:
module @SyncTensorsGraph.19 attributes {mhlo.cross_program_prefetches = [], mhlo.input_output_alias = [], mhlo.is_dynamic = false, mhlo.use_auto_spmd_partitioning = false} {
  func.func @main() -> (tensor<f32>, tensor<4x8x18x128xbf16>, tensor<4x8x18x128xbf16>, tensor<4x18x151936xbf16>) {
    %0 = sdy.constant dense<1.000000e+00> : tensor<f32> loc(#loc1)
    %1 = sdy.constant dense<0.000000e+00> : tensor<bf16> loc(#loc)
    %2 = stablehlo.broadcast_in_dim %1, dims = [] : (tensor<bf16>) -> tensor<4x8x18x128xbf16> loc(#loc)
    %3 = sdy.constant dense<0.000000e+00> : tensor<bf16> loc(#loc2)
    %4 = stablehlo.broadcast_in_dim %3, dims = [] : (tensor<bf16>) -> tensor<4x18x151936xbf16> loc(#loc2)
    return %0, %2, %2, %4 : tensor<f32>, tensor<4x8x18x128xbf16>, tensor<4x8x18x128xbf16>, tensor<4x18x151936xbf16> loc(#loc)
  } loc(#loc)
} loc(#loc)
#loc = loc(unknown)
#loc1 = loc("constant.1")
#loc2 = loc("broadcast.17")
```

**Comment by jameszianxuTT** [comment]
[Logs from main](https://raw.githubusercontent.com/jameszianxuTT/logdump/refs/heads/main/qt_small_main_pass_2026-04-23_18-27-29.log). Without AOTAutograd, we don't get this weird grad initialization graph without a mesh -> implied mesh shape 1x1 -> mesh reshape to 1x1 -> segfault on transferring tensors to host. 

(1) we shouldn't segfault on transferring tensors to host under any circumstances, that is an issue
(2) if we can get rid of the grad init graph or get it to run in the 1x2 mesh (idk how it bypasses that) then we might be able to get around this crash

Also, I wonder if a simpler training graph can repro this issue, running some tests. 

Anyways, I think this is part of a pattern of many failures we've seen recently and randomly around returning tensors to host on after PjrtTensor implementation. I don't think anyone's pinned down a root cause or a simple repro yet.

**Comment by jameszianxuTT** [comment]
Another thing I've run into is a corruption issue in metal kernel cache where, when pulling devices back to hosts, the generated kernel is cached in a corrupted path string like: 
```
-rw-rw-r--  1 jameszianxu jameszianxu        0 Apr 23 19:52 '~q'$'\317''K'$'\373\177'
drwxrwxr-x  2 jameszianxu jameszianxu        6 Apr 23 20:13 ''$'\246''g'$'\310''?'$'\373\177'

from the following:
2026-04-23 19:52:57.902 | critical |          Always | TT_THROW: Failed to generate binaries for writer_unary_interleaved_start_id filesystem error: cannot rename: No such file or directory [/home/jameszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/writer_unary_interleaved_start_id/7756022787363217060/brisc/.dephash] [/home/jameszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/writer_unary_interleaved_start_id/7756022787363217060/brisc/.dephash] (assert.hpp:104)
[Detaching after vfork from child process 266255]
sh: 1: cd: can't cd to ��K�
2026-04-23 19:52:57.974 | error    |    BuildKernels | trisc1 link failure -- cmd: cd ��K�meszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/tilize/17109946792624430764/trisc1/ && /opt/tenstorrent/sfpi/compiler/bin/riscv-tt-elf-g++ -O3 -Wl,--just-symbols=�1�K�6$��lianxu/aot/tt-xla/third_party/tt-mlir/src/tt-mlir/third_party/tt-metal/src/tt-metal/tt_metal/pre-compiled/8910724491701325957/trisc1/trisc1_weakened.elf �d�K�6$��l=auto -ffast-math -fno-exceptions -g -Wl,-z,max-page-size=16 -Wl,-z,common-page-size=16 -nostartfiles -mcpu=tt-wh-tensix -T/localdev/jameszianxu/aot/tt-xla/third_party/tt-mlir/src/tt-mlir/third_party/tt-metal/src/tt-metal/runtime/hw/toolchain/wormhole/kernel_trisc1.ld -Wl,--emit-relocs �K�6$��lianxu/aot/tt-xla/third_party/tt-mlir/src/tt-mlir/third_party/tt-metal/src/tt-metal/runtime/hw/lib/wormhole/substitutes.o /home/jameszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/tilize/17109946792624430764/trisc1/ -o ��K�meszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/tilize/17109946792624430764/trisc1/trisc1.14101125417909317745.elf (build.cpp:55)
2026-04-23 19:52:57.974 | critical |          Always | TT_THROW: Failed to open link failure log file ��K�meszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/tilize/17109946792624430764/trisc1/trisc1.elf.14101125417909317745.log (assert.hpp:104)
2026-04-23 19:52:58.001 | error    |    BuildKernels | Failed to rename temporary file ��K�meszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/tilize/17109946792624430764/trisc1/trisc1.elf.14101125417909317745.log to target file ��K�meszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/tilize/17109946792624430764/trisc1/trisc1.elf.log: No such file or directory (jit_build_utils.cpp:82)
2026-04-23 19:52:58.001 | error    |    BuildKernels | Failed to rename temporary file ��K�meszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/tilize/17109946792624430764/trisc1/trisc1.14101125417909317745.elf to target file ��K�meszianxu/.cache/tt-metal-cache/8910724491701325957/kernels/tilize/17109946792624430764/trisc1/trisc1.elf: No such file or directory (jit_build_utils.cpp:82)
2026-04-23 19:52:58.001 | critical |          Always | TT_THROW: Failed to generate binaries for tilize TT_THROW @ /localdev/jameszianxu/aot/tt-xla/third_party/tt-mlir/src/tt-mlir/third_party/tt-metal/src/tt-metal/tt_metal/jit_build/build.cpp:61: tt::exception
```

@jazpurTT Hit this a week ago doing something different, though induced the failure in the same way while materializing tensors to host on mesh shape change. His conclusions were that it is not easy to reproduce and we didn't get any feedback from metal on this cache corruption issue

The w/a for that was to avoid the mesh device reshape and the subsequent pulling of tensors to host. Still tbd is:
1. Can we avoid making this subgraph to initialize gradients or fold it into the bwd pass. it seems like it's specific to AOTAutograd per my previous comment
2. How did this subgraph get generated and intercepted by lazytensor but not have sharding annotations attached to it, since it was executed in an SPMD context?
3. Why are we seeing the various failures pulling tensors back to host? Is that related to (4)? (Maybe some runtime hacks can help here, eg. the double buffering of inputs we had before)
4. Why are we getting metal kernel cache corruption issues pulling tensors back to host?

**Comment by jameszianxuTT** [comment]
Answer for (2), as discovered by @pglusacTT 
`How did this subgraph get generated and intercepted by lazytensor but not have sharding annotations attached to it, since it was executed in an SPMD context?`

These graphs are a bit degenerate and have **no** inputs. We attach and infer sharding/spmd context from the inputs; so this is not an escape of SPMD but a gap in how we express and detect sharding/mesh.
`func.func @main() -> (tensor<f32>, tensor<4x8x18x128xbf16>, tensor<4x8x18x128xbf16>, tensor<4x18x151936xbf16>) {`

**Comment by acolicTT** [comment]
It should be solved: https://github.com/tenstorrent/tt-xla/pull/4439
